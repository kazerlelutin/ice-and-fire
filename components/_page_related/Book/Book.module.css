.container {
    display: grid;
    flex-direction: column;
    gap: 15px;
    padding: 10px;
}