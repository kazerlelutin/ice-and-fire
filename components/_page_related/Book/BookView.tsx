import classes from './Book.module.css'
import CommonLayout from '../../_layouts/CommonLayout/CommonLayout'
import { IBook } from '../../../types/book.interface'
import Loader from '../../_UI/Loader/Loader'
import BookDetails from '../../BookDetails/BookDetails'
import Characters from '../../Characters/Characters';

interface Props {
  readonly book: IBook | null
  readonly loading: boolean
}

export default function BookView({ book, loading }: Props) {
  return (
    <CommonLayout>
      <div className={classes.container}>
        {loading && <Loader />}
        {book && <>
          <BookDetails book={book} />
          <Characters CharacterIds={book.characters}/>
          </>}
      </div>
    </CommonLayout>
  )
}
