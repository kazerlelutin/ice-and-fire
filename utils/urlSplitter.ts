//for extract id 
export default function urlSplitter(url:string):string {
  const splitUrl = url.split('/')
  return splitUrl.at(-1)
}