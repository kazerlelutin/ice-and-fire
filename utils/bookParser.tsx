import { IapiBook } from "../types/apiBook.interface";
import { IBook } from "../types/book.interface";
import urlSplitter from "./urlSplitter";

export default function bookParser(book:IapiBook):IBook{

  return {
    id:urlSplitter(book.url),
    isbn: book.isbn,
    name: book.name,
    released: book.released,
    numberOfPages: book.numberOfPages,
    //merge all type of characters, just IDs
    characters: [...book.characters, ...book.povCharacters].reduce(
      (previousValue, currentValue) => {
        const 
        characterId = urlSplitter(currentValue),
        isExist = previousValue.find((id) => id === characterId)
        if (!isExist) previousValue.push(characterId)
        return previousValue
      },
      []
    ),
  }
}