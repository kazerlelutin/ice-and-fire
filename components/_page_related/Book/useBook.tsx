import { got } from '../../../datasources/got';
import { IapiBook } from '../../../types/apiBook.interface';
import bookParser from '../../../utils/bookParser';
import useFetch from '../../_hooks/useFetch';
interface Props {
  readonly id: string
}
export default function useBook({id}:Props){
  const {data:book,loading} = useFetch<IapiBook>(got,'books/'+id)

  return {book:book ? bookParser(book):null,loading}
}