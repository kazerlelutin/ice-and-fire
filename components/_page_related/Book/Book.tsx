import Factory from '../../_layouts/Factory/Factory';
import BookView from './BookView';
import useBook from './useBook';

const Books = Factory(useBook,BookView)
export default Books