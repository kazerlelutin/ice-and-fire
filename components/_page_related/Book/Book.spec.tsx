import { afterEach, describe, expect, it } from 'vitest'
import {cleanup } from '@testing-library/react'
import Book from './Book'
import { render} from '../../../utils/test-utils'

describe('<Book />', () => {
  afterEach(cleanup)

  it('Check snapshot', async () => {
    const { asFragment } = render(<Book/>)
    expect(asFragment()).toMatchSnapshot()
  })
})


