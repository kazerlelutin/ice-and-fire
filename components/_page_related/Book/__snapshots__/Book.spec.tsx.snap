// Vitest Snapshot v1

exports[`<Book /> > Check snapshot 1`] = `
<DocumentFragment>
  <div
    class="_container_f3f7d0"
  >
    <header
      class="_header_f3f7d0"
    >
      <div
        class="_container_46c4bf"
      >
        <div
          class="_container_5bb206"
        >
          Ice & Fire
        </div>
      </div>
    </header>
    <main
      class="_main_f3f7d0"
    >
      <div
        class="_container_596068"
      >
        books details, characters
      </div>
    </main>
  </div>
</DocumentFragment>
`;
